### PR TITLE
Disable unregister entity button in catalog if unauthorized

### PR DESCRIPTION
**Pull request by Joonpark13**
## Hey, I just made a Pull Request!

Most of the changes are as-is from the PRFC branch. Added docs, changesets, and tests.

image

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: plugins/catalog/src/components/EntityContextMenu/EntityContextMenu.test.tsx
Comment (mtlewis) [inline plugins/catalog/src/components/EntityContextMenu/EntityContextMenu.test.tsx:23]: This feels like a very targeted bit of mocking which might make the tests a little more brittle in the future. Did you explore the possibility of providing the MockPermissionApi in the test app used in this suite instead?

File: plugins/catalog/src/components/EntityLayout/EntityLayout.test.tsx
Comment (mtlewis) [inline plugins/catalog/src/components/EntityLayout/EntityLayout.test.tsx:39]: Ditto re. mocking - in this case we even have the api registry broken out for us already :tada:.

**Review by mtlewis — APPROVED** [review]
Looking great! Big chunk of the PRFC branch code heading for master too :tada:. Just one small suggestion around testing, but LGTM either way.

File: plugins/catalog/src/components/EntityContextMenu/EntityContextMenu.test.tsx
Comment (Joonpark13) [reply plugins/catalog/src/components/EntityContextMenu/EntityContextMenu.test.tsx]: Good catch, fixed!

File: plugins/catalog-react/src/hooks/useEntityPermission.ts
Comment (vinzscam) [inline plugins/catalog-react/src/hooks/useEntityPermission.ts:32]: this will make the component crash in case the entity is loading

Comment (joeporpeglia) [reply plugins/catalog-react/src/hooks/useEntityPermission.ts:32]: Seems like this works because the context menu isn't rendered until the entity is loaded, but this hook should probably handle the loading state to avoid misleading behavior in other use cases.

File: plugins/catalog-react/src/hooks/useEntityPermission.ts
Comment (joeporpeglia) [inline plugins/catalog-react/src/hooks/useEntityPermission.ts:34]: Requiring the full entity as input here will block the permission request until the entity is loaded. This seems fine for the "unregister component" UX because it's hidden behind a menu — just calling out that other views may want to optimize the request flow by loading the entity + permissions in parallel. Could add a comment about using `usePermission` directly in those cases.

**Review by joeporpeglia — COMMENTED** [review]
Looks good!

File: plugins/catalog-react/src/hooks/useEntityPermission.ts
Comment (Joonpark13) [reply plugins/catalog-react/src/hooks/useEntityPermission.ts]: I find the type casting in the implementation of `useEntity` a bit strange because it's casting `undefined` as `T`, so the return signature just looks like `entity` will always be `Entity`, whereas in reality it should be `Entity | undefined`. But it's a good point in terms of handling the loading case. Due to the no-conditional-calls rule of hooks I've had to make the hook a bit more complicated, but it should handle all cases now.

File: plugins/catalog-react/src/hooks/useEntityPermission.ts
Comment (Joonpark13) [reply plugins/catalog-react/src/hooks/useEntityPermission.ts]: Done.

File: plugins/catalog-react/src/hooks/useEntityPermission.test.tsx
Comment (Joonpark13) [inline plugins/catalog-react/src/hooks/useEntityPermission.test.tsx:32]: The most recent refactor of the hook meant moving away from using the provided mock of the api, in order to tune the return values of the underlying hooks to test each specific case in isolation.

Comment (freben) [reply plugins/catalog-react/src/hooks/useEntityPermission.test.tsx:32]: Is this really necessary, or only because you wanted to shorten the test setup code? I'd have preferred the longer and more explicit setup personally :)

**Review by freben — APPROVED** [review]
Great stuff :) Let's wait until tomorrow to merge tho so Rugvip kan look at it too

File: plugins/catalog-react/src/hooks/useEntityPermission.ts
Comment (joeporpeglia) [reply plugins/catalog-react/src/hooks/useEntityPermission.ts]: Oh yea that is strange! Didn't notice that before.. 🤔

**Review by joeporpeglia — APPROVED** [review]

**Review by Rugvip — APPROVED** [review]

**Pull request merged**
